mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -1930,6 +1956,96 @@ struct AttentionRewritePattern : public OpRewritePattern<tosa::MatMulOp> {
     return failure();
   }
 
+  // Result struct for sequence length mask detection
+  struct SeqLenMaskResult {
+    Value inputToContinue; // The value to continue pattern matching with
+    Value seqLen;          // The sequence length or prefix offset value
+    Value prefixOffset;    // The prefix offset value
+  };
+
+  // Helper to try detecting prefix causal pattern: add(row_indices, offset)
+  // Returns the offset value if successful
+  FailureOr<Value>
+  tryPrefixCausalPattern(Value input,
+                         const DenseSet<StringRef> &expandAndCollapse,
```

Comment:
nit: if it is known to be expanAndCollpase, move it's definition inside the function instead of passing as func arg.